### PR TITLE
✨ Feature ➾ Implement `taq instantiate-account` and `taq fund` to Generate and Fund Declared Accounts in a Target Environment

### LINKED ISSUE #1306: 🚧 Dev ➾ Accounts and funding MVP iteration 2
@### 🔗 Epic|Feature Link

_No response_

### 🚥 Definition of Done

- [x] 🎯 Planning and Design
- [x] 🏎️ Dev Implementation
- [x] 🏁 Test Plan Executed
- [x] 🛩️ Documentation Updated
- [ ] 🏟️ Peer / User Reviewed
- [ ] 🏆 Validated and Accepted

### ✅ To Do

_No response_

### 🔭 Scope
 
 🙅‍♂️ Out of Scope: 
- Default accounts
- Unrelated config.json refactoring
- No schema changes

### 🧰 Work Description

Extending the functionality of account funding to support the creation of declared accounts in a testnet environment, and adding a means to fund those accounts via transfers from the root account

This will likely involve the creation of two tasks for the Taquito plugin, one to handle the instantiation of accounts, and the other to handle transferring funds

### ⚖️ Acceptance Criteria

✨ Features: 
- [x] A task to generate accounts (instantiate the declared accounts in the current or targeted env)
- [x] A task to fund those accounts (transfer from root account)
- [x] Taquito plugin tasks have an optional flag to specify a source account (--sender) 
- [x] `taq fund` funds each declared account with the amount of Tez defined

Non-Functional:
- [x] Use batch operation to send operations to fund accounts
- [x] These tasks must fail gracefully if the user targets an unsupported env (ie sandbox)

### 🦄 Sizing

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by jchenche**
# 🌮 Taqueria PR

Fixes #1306 

## 🪁 Description

Introduced the `taq instantiate-account` task to instantiate declared accounts and the `taq fund` task to fund instantiated accounts. Both tasks work under a specific environment.

Additionally, users can specify the sender of operations via the `--sender` flag.

Changed `--tez` to `--mutez` for `taq transfer` to allow more granular amount. And give origination the ability to specify an initial balance for the originated contract.

## 🪂 Pre-Merge Checklist (Definition of Done)

🚦 Required to merge:

- [x] ⛱️ I have completed this PR template in full and updated the title appropriately
- [x] ⛵ My code builds cleanly, and I have manually tested the changes
- [ ] 🏄‍♂️ Another team member has built this branch and done manual testing on the change
- [x] 🏖️ New and existing unit tests pass locally and in CI
- [x] 🐬 I have commented my code, particularly in hard-to-understand areas
- [x] 🤿 Corresponding changes have been made to all documentation

## 🛸 Type of Change

- [ ] 🧽 Chore ➾
- [ ] 🛠️ Fix ➾
- [x] ✨ Feature ➾
- [x] 👷 Refactor ➾
- [ ] 🧪 Pre-Release ➾
- [ ] 🚀 Release ➾

### CODE REVIEW

File: taqueria-plugin-taquito/fund.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+				const alias = instantiatedAccount[0];
+				const aliasInfo = instantiatedAccount[1];
+
+				const declaredMutez: number | undefined = getDeclaredAccounts(parsedArgs)[alias];
+				const currentBalanceInMutez = (await tezos.tz.getBalance(aliasInfo.publicKeyHash)).toNumber();
+				const amountToFillInMutez = declaredMutez ? Math.max(declaredMutez - currentBalanceInMutez, 0) : 0;
+
+				if (!declaredMutez) {
+					sendWarn(
+						`Warning: ${alias} is instantiated in the target environment but not declared in the root level "accounts" field of ./.taq/config.json so ${alias} will not be funded as we don't have a declared tez amount set there for ${alias}\n`,
```
Comment (hu3man) [inline taqueria-plugin-taquito/fund.ts:41]: Technical writing should be in second person narrative, so avoid terms like "we", "us", etc

In second person, it's mostly about using "you"

File: taqueria-plugin-taquito/fund.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+		const [envType, nodeConfig] = await getEnvTypeAndNodeConfig(parsedArgs, env);
+		if (envType !== 'Network') return sendAsyncErr('taq fund can only be executed in a network environment');
+		const tezos = await configureToolKitForNetwork(parsedArgs, nodeConfig);
+
+		const instantiatedAccounts = getNetworkInstantiatedAccounts(nodeConfig);
+
+		const accountsInfo = await getAccountsInfo(parsedArgs, tezos, instantiatedAccounts);
+		if (accountsInfo.length === 0) {
+			return sendJsonRes(
+				`All instantiated accounts in the current environment, "${parsedArgs.env}", are funded up to or beyond the declared amount.`,
```
Comment (hu3man) [inline taqueria-plugin-taquito/fund.ts:79]: Trailing period

File: taqueria-plugin-taquito/instantiate_account.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+					`Note: ${declaredAccountAlias} is already instantiated in the current environment, "${parsedArgs.env}".`,
+				);
+			}
+		}
+
+		if (accountsInstantiated.length !== 0) {
+			return sendJsonRes(
+				`Accounts instantiated: ${
+					accountsInstantiated.join(', ')
+				}.\nPlease execute "taq fund" targeting the same environment to fund these accounts.`,
```
Comment (hu3man) [inline taqueria-plugin-taquito/instantiate_account.ts:38]: Trailing period

File: taqueria-plugin-taquito/instantiate_account.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+			);
+		} else {
+			return sendJsonRes(`No accounts were instantiated.`);
+		}
+	} catch {
+		return sendAsyncErr('No operations performed.');
```
Comment (hu3man) [inline taqueria-plugin-taquito/instantiate_account.ts:44]: Trailing period

**Review by hu3man — APPROVED** [review]
Awesome work @jchenche!

Code looks good to me, just left a few small comments about cli output and other text

File: taqueria-plugin-taquito/common.ts [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+): Promise<void> => {
+	const tezos = new TezosToolkit(network.rpcUrl as string);
+	const key = await getAccountPrivateKey(parsedArgs, network, account);
+	await importKey(tezos, key);
+};
+
+export const handleOpsError = (err: unknown, env: string): Promise<never> => {
+	if (err instanceof Error) {
+		const msg = err.message;
+		if (/ENOTFOUND/.test(msg)) return sendAsyncErr('The RPC URL may be invalid. Check ./.taq/config.json');
```
Comment (AlirezaHaghshenas) [inline taqueria-plugin-taquito/common.ts:181]: Just some thoughts:
Here, we swallow the error's message and return a friendly message base on matching a regex. This is all good unless there is a possible other error that matches this regex, then this behavior becomes misleading.

Maybe still returning the original error to the calling code can prevent that.

Comment (jchenche) [reply taqueria-plugin-taquito/common.ts:181]: @AlirezaHaghshenas The original error takes up the entire command screen 😞. That's why I opted for this instead. I tried to make the regexes precise enough such that it only captures the desired error. I still want to refine this part a little but I want to do that as part of the next iteration.

File: taqueria-plugin-taquito/instantiate_account.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+		}
+
+		if (accountsInstantiated.length !== 0) {
+			return sendJsonRes(
+				`Accounts instantiated: ${
+					accountsInstantiated.join(', ')
+				}.\nPlease execute "taq fund" targeting the same environment to fund these accounts`,
+			);
+		} else {
+			return sendJsonRes(`No accounts were instantiated`);
```
Comment (AlirezaHaghshenas) [inline taqueria-plugin-taquito/instantiate_account.ts:41]: Here, is it a good idea to tell the user why "No accounts were instantiated"?

File: taqueria-plugin-taquito/instantiate_account.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+		}
+	} catch {
+		return sendAsyncErr('No operations performed');
```
Comment (AlirezaHaghshenas) [inline taqueria-plugin-taquito/instantiate_account.ts:44]: And in here, include the original error?

**Review by AlirezaHaghshenas — APPROVED** [review]

File: taqueria-plugin-taquito/common.ts (hunk 5 above)
Comment (AlirezaHaghshenas) [reply taqueria-plugin-taquito/common.ts:181]: Yeah, I understand that sometimes for practical reasons (like not giving the user a heart attack) we have to make a compromise.

**Comment by AlirezaHaghshenas** [comment]
I love how code is getting simpler while it is doing more things than before. Good job Jimmy!

File: taqueria-plugin-taquito/instantiate_account.ts (hunk 6 above)
Comment (jchenche) [reply taqueria-plugin-taquito/instantiate_account.ts:41]: Will do!

File: taqueria-plugin-taquito/instantiate_account.ts (hunk 7 above)
Comment (jchenche) [reply taqueria-plugin-taquito/instantiate_account.ts:44]: What do you mean?

Comment (jchenche) [reply taqueria-plugin-taquito/instantiate_account.ts:44]: The errors before this error (`'No operations performed'`) will tell the user if it came from an `originate` task or `transfer` task if that's what you meant. This is more like to assure users that no operations were injected into the chain.

Comment (AlirezaHaghshenas) [reply taqueria-plugin-taquito/instantiate_account.ts:44]: I mean, here we catch any possible error, but do not include any information from them in the error message, so the user knows that "No operations performed" but not why.

Comment (jchenche) [reply taqueria-plugin-taquito/instantiate_account.ts:44]: Oh it's caught in the `handleOpsError` function

**Pull request merged**